src/main/java/com/conveyal/taui/controllers/AggregationAreaController.java

```diff
@@ -99,45 +112,69 @@ public static AggregationArea createAggregationArea (Request req, Response res)
-        Geometry merged = union.union();
+        List<SimpleFeature> features = reader.wgs84Stream().collect(Collectors.toList());
+
+        Map<String, Geometry> areas = new HashMap<>();
+
+        if (Boolean.parseBoolean(req.params("union"))) {
+            List<Geometry> geometries = features.stream().map(f -> (Geometry) f.getDefaultGeometry()).collect(Collectors.toList());
+            UnaryUnionOp union = new UnaryUnionOp(geometries);
+            // Use provided name directly for this single unioned feature
+            areas.put(maskName, union.union());
+        } else {
+            // Use provided name to read a name for each feature
+            try {
+                features.stream().forEach(f -> areas.put(f.getProperty(maskName).getValue().toString(), (Geometry) f.getDefaultGeometry()));
+            } catch (NullPointerException e) {
+                throw new AnalysisServerException("The supplied name was not a property of the uploaded features. " +
+                        "Double check that the name corresponds to a shapefile column.");
+            }
+        }
+
+        areas.forEach((String name, Geometry geometry) -> {
+            Envelope env = geometry.getEnvelopeInternal();
+            Grid maskGrid = new Grid(SeamlessCensusGridExtractor.ZOOM, env.getMaxY(), env.getMaxX(), env.getMinY(), env.getMinX());
+
+            // Store the percentage each cell overlaps the mask, scaled as 0 to 100,000
+            List<Grid.PixelWeight> weights = maskGrid.getPixelWeights(geometry, true);
+            weights.forEach(pixel -> maskGrid.grid[pixel.x][pixel.y] = pixel.weight * 100_000);
 
-        Envelope env = merged.getEnvelopeInternal();
-        Grid maskGrid = new Grid(SeamlessCensusGridExtractor.ZOOM, env.getMaxY(), env.getMaxX(), env.getMinY(), env.getMinX());
+            ObjectMetadata metadata = new ObjectMetadata();
+            metadata.setContentEncoding("gzip");
+            metadata.setContentType("application/octet-stream");
 
-        // Store the percentage each cell overlaps the mask, scaled as 0 to 100,000
-        List<Grid.PixelWeight> weights = maskGrid.getPixelWeights(merged, true);
-        weights.forEach(pixel -> maskGrid.grid[pixel.x][pixel.y] = pixel.weight * 100_000);
+            AggregationArea aggregationArea = new AggregationArea();
+            aggregationArea.name = name;
+            aggregationArea.regionId = regionId;
 
-        ObjectMetadata metadata = new ObjectMetadata();
-        metadata.setContentEncoding("gzip");
-        metadata.setContentType("application/octet-stream");
+            // Set `createdBy` and `accessGroup`
+            aggregationArea.accessGroup = req.attribute("accessGroup");
+            aggregationArea.createdBy = req.attribute("email");
 
-        AggregationArea aggregationArea = new AggregationArea();
-        aggregationArea.name = maskName;
-        aggregationArea.regionId = regionId;
+            File gridFile = new File(tempDir, "weights.grid");
```

Comment:
Rather than creating a temp dir, always containing a single file of the same name, we should probably just create a single temp file with a system-generated name.

Reply:
The temp dir is used above to hold the uploaded shapefile components, but it may not be needed when we move to creating aggregation areas from the resource pool 👍 

Would you suggest a UUID prefix as the system-generated name?